File: Python/vdFile.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
@author: Florian Timm
@version: 2017.11.27
"""

import datetime
from abc import abstractmethod, ABC

from vdPoint import VdPoint


class VdFile(ABC):

    """ abstract class for saving data """

    def __init__(self, conf, filename=""):
        """
        Creates a new ascii-file
        :param conf: configuration file
        :type conf: configparser.ConfigParser
        :param filename: name and path to new file
        :type filename: str
        """
        self.__conf = conf
        # create file
        self.__writing_queue = []
        self._open(filename)

    def __get_writing_queue(self):
        """
        Returns points in queue
        :return: points in queue
        :rtype: VdPoint[]
        """
        return self.__writing_queue

    writing_queue = property(__get_writing_queue)

    def clear_writing_queue(self):
        """ clears writing queue """
        self.__writing_queue = []

    def _make_filename(self, file_format, file_name=""):
        """
        generates a new file_name from timestamp
        :param file_format: file suffix
        :type file_format: str
        :return: string with date and suffix
        :rtype: str
        """
        if file_name == "":
            name = self.__conf.get("file", "namePre")
            name += datetime.datetime.now().strftime(
                self.__conf.get("file", "timeFormat"))
            name = "." + file_format
            return name
        elif not file_name.endswith("." + file_format):
            return file_name + "." + file_format
        return file_name

    def write_data(self, data):
        """
        adds data and writes it to file
        :param data: ascii data to write
        :type data: VdPoint[]
        """
        self.add_dataset(data)
        self.write()

    def add_point(self, p):
        """
        Adds a point to write queue
        :param p: point
        :type p: VdPoint
        """
        self.__writing_queue.append(p)

    def add_dataset(self, dataset):
        """
        adds multiple points to write queue
        :param dataset: multiple points
        :type dataset: VdPoint[]
        """
        self.__writing_queue.extend(dataset)

    def read_from_txt_file(self, filename, write=False):
        """
        Parses data from txt file
        :param filename: path and filename of txt file
        :type filename: str
        :param write: write data to new file while reading txt
        :type write: bool
        """
        txt = open(filename)

        for no, line in enumerate(txt):
            try:
                p = VdPoint.parse_string(self.__conf, line)
                self.writing_queue.append(p)
                # print("Line {0} was parsed".format(no + 1))
            except ValueError as e:
                print("Error in line {0}: {1}".format(no + 1, e))

            if write and len(self.writing_queue) > 50000:
                self.write()
        if write:
            self.write()

    @abstractmethod
    def write(self):
        """ abstract method: should write data from writing_queue """
        pass

    @abstractmethod
    def _open(self, filename):
        """
        abstract method: should open file for writing
        :param filename: file name
        :type filename: str
        """
        pass

    @abstractmethod
    def close(self):
        """ abstract method: should close file """
        pass

File: C++/src/VdFile.cpp
/*
 * VdFile.cpp
 *
 *  Created on: 26.11.2017
 *      Author: timm
 */
#include <string>
#include <iostream>
#include <list>
using namespace std;
#include "VdFile.h"

list<VdPoint>& VdFile::getWritingQueue() {
	/*
	 Returns points in queue
	 :return: points in queue
	 :rtype: VdPoint[]
	 */
	return this->writingQueue;
}

void VdFile::clearWritingQueue() {
	// clears writing queue
	this->writingQueue.clear();
}

inline bool ends_with(string const & value, string const & ending) {
	if (ending.size() > value.size())
		return false;
	return equal(ending.rbegin(), ending.rend(), value.rbegin());
}

string VdFile::makeFilename(string file_format, string file_name) {
	/*
	 generates a new file_name from timestamp
	 :param file_format: file suffix
	 :type file_format: str
	 :return: string with date and suffix
	 :rtype: str
	 */
	string file_ending = string(".") + file_format;
	if (file_name.empty()) {
		return (string("noname") + file_ending);
	} else if (ends_with(file_name, file_ending)) {
		return file_name;
	}
	return file_name + file_ending;
}

void VdFile::writeData(list<VdPoint>* dataset) {
	/*
	 adds data and writes it to file
	 :param data: ascii data to write
	 :type data: VdPoint[]
	 */
	this->addDataset(dataset);
	this->write();
}

void VdFile::addPoint(VdPoint point) {
	/*
	 Adds a point to write queue
	 :param p: point
	 :type p: VdPoint
	 */
	this->writingQueue.push_back(point);
}

void VdFile::addDataset(list<VdPoint>* dataset) {
	/*
	 adds multiple points to write queue
	 :param dataset: multiple points
	 :type dataset: VdPoint[]
	 */
	cout << dataset->size() << " datasets added" << endl;

	VdPoint p (1,2,3,4,5);

	//this->writingQueue.push_back(p);

	this->writingQueue.insert(this->writingQueue.begin(), dataset->begin(), dataset->end());

	cout << this->writingQueue.size();
	cout << " datasets" << endl;
}

